Issue 53 retrace

pelssers opened the pull request:
~~_Needs to be merged after #54 , might need a rebase_~~ Done, rebased onto master

This implements retracing of the loglikelihood function (and loglikelihood_grad2) whenever `LogLikelihood.set_data` is called, solving issue #53 

A test has been added to test exactly this issue.

Ideally we would rewrite loglikelihood to accept a `datatensor` and `ptensor` just like the `differential_rate`. Unfortunately we can't write an input signature at the moment since they can't take kwargs.

Edit:
I've added a commit which makes tracing of the likelihood via `set_data` optional (by adding the `autograph` flag to `set_data`). Also when initializing `LogLikelihood` tracing will only be done if data is supplied. This ensures we don't waste time by tracing if we're only calling `log_likelihood` with `autograph=False`.
Since tracing is optional I've added a flag to check whether a retrace has been made after the most recent `set_data` call.

pelssers commented on flamedisx/likelihood.py:
```diff
@@ -62,6 +60,10 @@ def __init__(
+        # Keep track of whether we traced ll after setting data
+        self.traced_after_set_data = False
+        # Only trace when we have data
+        trace_ll = False if data is None else True
```
We can't do that since we `del data` somewhere during the init, so we'd have to go into sources and check, but then there might be different datasets for different source etc. This seems to me the cleanest solution.

JelleAalbers replied:
Thanks, I missed that. Agree this makes sense then. We could remove the `del data` but that might cause other problems later (as the comment notes we should only use the data in the sources after it's set, since that has the correct annotations).

pelssers replied:
I agree that the likelihood init is getting very long, we could clean it up by initializing the sources with no data and then call `set_data` at the end (similar to the Source init), then it will deal with updating the padding, batches and also trigger the tracing.
Feel free to implement or open an issue, otherwise  let's leave it as is.

pelssers commented on flamedisx/likelihood.py:
```diff
@@ -148,13 +150,23 @@ def __init__(
+            self.trace_log_likelihood()
+            self.trace_log_likelihood_grad2()
+
+            self.traced_after_set_data = True
```
Good idea, will do

pelssers commented:
Thanks, I've implemented your suggestions. I've also added one more commit removing an assert in source.py with is no longer necessary after improving the padding in #54 
Performance wise you can choose to either retrace after every `set_data` and use the faster ll evaluation, or don't retrace and use eager mode. At least both options are now available to users, I guess it now depends on the complexity of the model which is the faster approach.
